### PR TITLE
Updated CI documentation.

### PR DESCRIPTION
**Pull request by ndmckinley**
Update the CI documentation for the magician v2 - mostly a revert-of-the-revert of my first attempt at updating the magician.

(don't pay attention to the diff noise - this is a full rewrite, it just uses just enough of the old words for the differ to be confused.)

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

### CODE REVIEW

File: .ci/README.md [hunk 1]
```diff
@@ -1,150 +1,47 @@
...
+
+## Design choices & tradeoffs
+* The downstreams share some setup code in common - especially TPG and TPGB.  We violated the DRY principle by writing separate workflows for each repo.  In practice, this has substantially reduced the amount of code - the coordination layer above the two repos was larger than the code saved by combining them.  We also increase speed, since each Action runs separately.
+* The downstream push doesn't wait for checks on its PRs against downstreams.  This may inconvenience some existing workflows which rely on the downstream PR checks.  This ensures that merge conflicts never come into play, since the downstreams never have dangling PRs, but it requires some up-front work to get those checks into the differ.
```
Comment (chrisst) [inline .ci/README.md]: This feels like an actionable statement but I don't know what action should be taken. Are you referring to things like travis? If so, I'm not sure what "it requires some up-front work to get those checks into the differ" means. Can you rephrase it in terms of what the state of the world is right now?

File: .ci/README.md [hunk 2]
```diff
@@ -1,150 +1,47 @@
...
+* The downstream push is disconnected from the output of the differ (but runs the same code).  This means that the diff which is approved isn't guaranteed to be applied *exactly*, if for instance magic modules' behavior changes on master between diff generation and downstream push.  This is also intended to avoid merge conflicts by, effectively, rebasing each commit on top of master before final generation is done.
+    * Imagine the following situation: PR A and PR B are opened simultaneously. PR A changes the copyright date in each file to 2020. PR B adds a new resource. PR A is merged seconds before PR B, so they are picked up in the same push-downstream run.  The commit from PR B will produce a new file with the 2020 copyright date, even though the diff said 2019, since PR A was merged first.
+* We deleted the submodules.  They weren't useful to us and they were annoying to update - they're not in use anymore as far as we know - but it's possible there's some long-forgotten workflow that someone is using which will be damaged.  So far, we haven't seen any such issue.
```
Comment (chrisst) [inline .ci/README.md]: ❤️

File: .ci/README.md [hunk 3]
```diff
@@ -1,150 +1,47 @@
...
 
-When merging PR #1, if we update `downstream-repo-a` to the merge commit for PR #7, we will go backwards, erasing PR #2.  If we update to `master`, we will definitely include both the changes of both PR #1 and PR #2.
+### Downstream Pushing
+The Magician maintains a set of tags called `$REPO-sync` that tracks the Magic Modules commit the downstreams are up to date with.
 
--------
+In effect, this means that each downstream commit will correspond 1:1 to an MM commit. If an MM commit had no changes in a downstream, no commit will be created.  We are enforcing squash-merges-only in Magic Modules.
 
-# Troubleshooting The Magician
+Downstream pushing can only take place if the sync tag points to the commit which precedes the commit that is being pushed.  It is possible for the pipeline to get stuck if a commit is merged into Magic Modules which cannot be generated - this happens most often when the Gemfile.lock is updated.  We expect the failure case to be somewhat uncommon.
```
Comment (chrisst) [inline .ci/README.md]: Gemfiles have been updated twice in the last month. Do you think it's worth documenting what the Gemfile update should look like?

File: .ci/README.md (hunk 1 above)
Comment (ndmckinley) [reply .ci/README.md]: Clarified!

File: .ci/README.md (hunk 3 above)
Comment (ndmckinley) [reply .ci/README.md]: Sure!
